app/services/amr/data_feed_upserter.rb

```diff
@@ -1,52 +1,28 @@
-require 'upsert'
-require 'upsert/active_record_upsert'
-require 'upsert/connection/postgresql'
-require 'upsert/connection/PG_Connection'
-require 'upsert/merge_function/PG_Connection'
-require 'upsert/column_definition/postgresql'
-
 module Amr
   class DataFeedUpserter
-    attr_reader :inserted_record_count
+    attr_reader :inserted_record_count, :upserted_record_count
 
     def initialize(array_of_data_feed_reading_hashes, amr_data_feed_import_log_id)
       @array_of_data_feed_reading_hashes = array_of_data_feed_reading_hashes
       @amr_data_feed_import_log_id = amr_data_feed_import_log_id
     end
 
     def perform
-      @inserted_record_count = upsert_records(@array_of_data_feed_reading_hashes)
+      add_import_log_id_and_dates_to_hash
+      result = AmrDataFeedReading.upsert_all(@array_of_data_feed_reading_hashes, unique_by: [:mpan_mprn, :reading_date])
+      result.rows.flatten.size
     end
 
   private
 
-    def upsert_records(array_of_data_feed_reading_hashes)
-      @existing_records = AmrDataFeedReading.count
-      Upsert.batch(AmrDataFeedReading.connection, AmrDataFeedReading.table_name) do |upsert|
-        array_of_data_feed_reading_hashes.each { |row| upsert_record(upsert, row) }
+    def add_import_log_id_and_dates_to_hash
+      created_at = DateTime.now.utc
+      updated_at = DateTime.now.utc
```

Comment:
We still have to add these manually - hopefully made better in a future Rails release